pkg/cri/server/sandbox_stats_windows.go

```diff
@@ -19,21 +19,324 @@ package server
+	req := &tasks.MetricsRequest{}
+	var containers []containerstore.Container
+	for _, cntr := range c.containerStore.List() {
+		if cntr.SandboxID != sandbox.ID {
+			continue
+		}
+		containers = append(containers, cntr)
+		req.Filters = append(req.Filters, "id=="+cntr.ID)
+	}
+
+	//add sandbox container as well
+	req.Filters = append(req.Filters, "id=="+sandbox.ID)
+
+	resp, err := c.client.TaskService().Metrics(ctx, req)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to fetch metrics for tasks: %w", err)
+	}
+	return resp.Metrics, containers, nil
+}
+
+func (c *criService) windowsCPUAndMemoryStats(stats *wstats.Statistics, oldStats *stats.ContainerStats) (*runtime.WindowsContainerStats, error) {
+	var cs runtime.WindowsContainerStats
+	if stats != nil {
+		wstats := stats.GetWindows()
+		if wstats == nil {
+			return nil, fmt.Errorf("windows stats is empty")
+		}
+		if wstats.Processor != nil {
+			cs.Cpu = &runtime.WindowsCpuUsage{
+				Timestamp:            wstats.Timestamp.UnixNano(),
+				UsageCoreNanoSeconds: &runtime.UInt64Value{Value: wstats.Processor.TotalRuntimeNS},
+			}
+		}
+
+		// Calculate NanoCores
+		if cs.Cpu.UsageCoreNanoSeconds != nil {
+			nanoCoreUsage := getUsageNanoCores(oldStats, cs.Cpu.UsageCoreNanoSeconds.Value, wstats.Timestamp)
+			cs.Cpu.UsageNanoCores = &runtime.UInt64Value{Value: nanoCoreUsage}
+		}
+
+		if wstats.Memory != nil {
+			cs.Memory = &runtime.WindowsMemoryUsage{
+				Timestamp: wstats.Timestamp.UnixNano(),
+				WorkingSetBytes: &runtime.UInt64Value{
+					Value: wstats.Memory.MemoryUsagePrivateWorkingSetBytes,
+				},
+			}
+		}
+
+	}
+	return &cs, nil
+}
+
+func windowsNetworkUsage(ctx context.Context, sandbox sandboxstore.Sandbox) *runtime.WindowsNetworkUsage {
+	eps, err := hcn.GetNamespaceEndpointIds(sandbox.NetNSPath)
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("unable to retrieve windows endpoint metrics for netNsPath: %v", sandbox.NetNSPath)
+		return nil
+	}
+	networkUsage := &runtime.WindowsNetworkUsage{
+		Timestamp: time.Now().UnixNano(),
+	}
+	for _, ep := range eps {
+		endpointStats, err := hcsshim.GetHNSEndpointStats(ep)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("unable to gather stats for endpoint: %s", ep)
+			continue
+		}
+		rtStats := runtime.WindowsNetworkInterfaceUsage{
+			Name:             endpointStats.EndpointID,
+			RxBytes:          &runtime.UInt64Value{Value: endpointStats.BytesReceived},
+			RxPacketsDropped: &runtime.UInt64Value{Value: endpointStats.DroppedPacketsIncoming},
+			TxBytes:          &runtime.UInt64Value{Value: endpointStats.BytesSent},
+			TxPacketsDropped: &runtime.UInt64Value{Value: endpointStats.DroppedPacketsOutgoing},
+		}
+		networkUsage.Interfaces = append(networkUsage.Interfaces, &rtStats)
+
+		// if the default interface isn't set add it.
+		// We don't have a way to determine the default interface in windows
```

Comment:
This doesn't seem like the right then to do on Windows but I am struggling to think of a better solution.
@daschott / @MikeZappa87 do you have any thoughts here?

Reply:
These are total counters of lifetime of the pod. What kind of stats are you looking for here? Is this for a K8s feature? What network pod stats are exposed by Linux K8s?

Reply:
This is for a [KEP ](https://github.com/kubernetes/enhancements/issues/2371)to move pod stats out of kubelet.  Kubelet will call the CRI endpoint instead of doing the work itself.  

For what it is worth, This is how it is done in kubelet today: https://github.com/kubernetes/kubernetes/blob/bb6edfb96b362736a66946ca7d88774aca167cb4/pkg/kubelet/stats/cri_stats_provider_windows.go#L91-L93  

I spoke with @katiewasnothere awhile back and we didn't find another way to do it.